### DISCUSSION TITLE
Implementation of clipping tool

### DISCUSSION THREAD

**Discussion by skc1712** [Kitware/trame#84] [Q&A]
Hi, 

I am trying to add a clipping tool as in Paraview for an example in Trame. For that, I am trying to follow this [example](https://kitware.github.io/vtk-examples/site/Cxx/Meshes/ClipClosedSurface/). As it is for the clipping of polygonal data, I am wondering how I can use it in the geometry of this [example](https://github.com/Kitware/trame-tutorial/blob/master/04_application/solution.py). Perhaps, if the geometry can be converted to polygonal data, I could continue as in the clipping example, but I am not sure if/how that can be done. 

Any comments are much appreciated. 

Best,

**Comment by skc1712** [comment]
Update: I have found [this](https://kitware.github.io/vtk-examples/site/Cxx/PolyData/GeometryFilter/) already helpful.

**Comment by jourdain** [comment]
Did you solved your problem on your own or do you still have a question?

**Comment by skc1712** [comment]
Hi,

Thank you. I could solve the problem partly, so if it's OK, I would use this thread to ask the other part as well. So far, I have been able to implement the clipping tool but just the intersection with the XZ-plane located at the origin as shown in the image here:

![Screenshot from 2022-07-06 13-41-31](https://user-images.githubusercontent.com/79941817/177544283-47ca648e-c1c8-49cc-89f3-f128cc1d0c2c.png)


My next aim is to include a drop-down menu where the local user can choose from the provided list of planes and visualise the results and then if possible, implement a slider which can be used to translate the initial plane along the normal axis. 


To obtain the image above, besides other little changes, following this [example](https://kitware.github.io/vtk-examples/site/Cxx/Meshes/ClipClosedSurface/), I have added the following snippet of the code. Do you have any idea what could be the right way to implement the abovementioned, perhaps to start with, the option of adding different choices of planes, i.e., YZ, XY-planes located at the origin?


`# --------------------------------------------------`
`# clip`
`geometryFilter = vtkGeometryFilter()          `
`geometryFilter.SetInputConnection(reader.GetOutputPort())   `
`geometryFilter.Update()          `
`polyData = geometryFilter.GetOutput()     `

`colors = vtkNamedColors()        `

`center = polyData.GetCenter()`

`plane1 = vtkPlane()     `
`plane2 = vtkPlane()      `
`plane3 = vtkPlane()     `
`# center = reader.GetOutput().GetCenter()  `
`plane1.SetOrigin(center[0], center[1], center[2])`
`plane1.SetNormal(0, 1, 0)`
`plane2.SetOrigin(center[0], center[1], center[2])`
`plane2.SetNormal(0, 0, 1)`
`plane3.SetOrigin(center[0], center[1], center[2])`
`plane3.SetNormal(1, 0, 0)`

`planes = vtkPlaneCollection()  `
`planes.AddItem(plane1)       `
`# planes.AddItem(plane2)       `
`# planes.AddItem(plane3)       `

`clipper = vtkClipClosedSurface()      `
`clipper.SetInputData(polyData)`

`clipper.SetClippingPlanes(planes)   `
`clipper.SetActivePlaneId(2)      `
`clipper.SetScalarModeToColors()     `

`clip_mapper = vtkDataSetMapper()  `
`clip_mapper.SetInputConnection(clipper.GetOutputPort())      `

`clip_actor = vtkActor()          `
`clip_actor.SetMapper(clip_mapper)     `
`# clip_actor.GetProperty().SetColor(1, 0.3882, 0.2784)   `
`clip_actor.GetProperty().SetInterpolationToFlat()        `

`renderer.AddActor(clip_actor)`

`clip_lut = clip_mapper.GetLookupTable()`
`clip_lut.SetHueRange(0.666, 0.0)`
`clip_lut.SetSaturationRange(1.0, 1.0)`
`clip_lut.SetValueRange(1.0, 1.0)`
`clip_lut.Build()`

`clip_mapper.SetLookupTable(clip_lut)             `

`# Mesh: Color by default array`
`clip_mapper.SelectColorArray(default_array.get("text"))`
`clip_mapper.GetLookupTable().SetRange(default_min, default_max)`

`if default_array.get("type") == vtkDataObject.FIELD_ASSOCIATION_POINTS:`
`    clip_mapper.SetScalarModeToUsePointFieldData()`
`else:`
`    clip_mapper.SetScalarModeToUseCellFieldData()`

`clip_mapper.SetScalarVisibility(True)`
`clip_mapper.SetUseLookupTableScalarRange(True)`

Thank you,

**Reply by jourdain** [reply]
Just use 1 plane and update the normal on that plane. No need to have 3 instances of planes.

**Comment by skc1712** [comment]
Apparently, what I tried above only clips the geometry and doesn't display the variables on that geometry. So, if there is any way of doing that, it would be helpful to know.

**Reply by jourdain** [reply]
You can refer to [VTK example](https://kitware.github.io/vtk-examples/site/) on how to use VTK.

**Comment by indominus-hub** [comment]
I am also interested on this feature, but I was unable to find a suitable example on VTK (like, for example, Colour a clip by Temperature)

Could you guys suggest an example where a variable is displayed on a clip/slice?

Many thanks in advance

**Reply by jourdain** [reply]
Would those be good examples?
- https://kitware.github.io/vtk-examples/site/Python/VisualizationAlgorithms/CutStructuredGrid/
- https://kitware.github.io/vtk-examples/site/Python/VisualizationAlgorithms/WarpCombustor/

Maybe the main issue with those examples is that they rely on the default scalars where in your case you aim to select it yourself.

But that example provide specific code call for coloring by a specific field: https://kitware.github.io/vtk-examples/site/Cxx/VisualizationAlgorithms/ColorIsosurface/

**Comment by indominus-hub** [comment]
Hi Jourdain,

Thanks for your reply.

Actually CutStructuredGrid looks nice, but the point is that we are working with vtu. Should this cutter work also on unstructured data?

The idea would be to create a user defined clip (or if easier, a slice), and then colour the resulting object by variable (velocity, pressure...or any other result available in the vtu file)

**Comment by jourdain** [comment]
The vtkCutter can work on any vtkDataSet which include vtkUnstructuredGrid.
